tests/utils/span.py

```diff
@@ -0,0 +1,382 @@
+        assert len(self.spans) == 0, 'Span count {0}'.format(len(self.spans))
+
+    def filter_spans(self, *args, **kwargs):
+        """
+        Helper to filter current spans by provided parameters.
+
+        This function will yield all spans whose `TestSpan.matches` function return `True`
+
+        :param *args: Positional arguments to pass to :meth:`tests.utils.span.TestSpan.matches`
+        :type *args: list
+        :param *kwargs: Keyword arguments to pass to :meth:`tests.utils.span.TestSpan.matches`
+        :type **kwargs: dict
+        :returns: generator for the matched :class:`tests.utils.span.TestSpan`
+        :rtype: generator
+        """
+        for span in self.spans:
+            # ensure we have a TestSpan
+            if not isinstance(span, TestSpan):
+                span = TestSpan(span)
+
+            if span.matches(*args, **kwargs):
+                yield span
+
+    def find_span(self, *args, **kwargs):
+        """
+        Find a single span matches the provided filter parameters.
+
+        This function will find the first spans whose `TestSpan.matches` function return `True`
+
+        :param *args: Positional arguments to pass to :meth:`tests.utils.span.TestSpan.matches`
+        :type *args: list
+        :param *kwargs: Keyword arguments to pass to :meth:`tests.utils.span.TestSpan.matches`
+        :type **kwargs: dict
+        :returns: The first matching span
+        :rtype: :class:`tests.utils.span.TestSpan`
+        """
+        span = next(self.filter_spans(*args, **kwargs), None)
+        assert span is not None, (
+            'No span found for filter {0!r} {1!r}, have {2} spans'
+            .format(args, kwargs, len(self.spans))
+        )
+        return span
+
+
+
+class TestSpanNode(TestSpan, TestSpanContainer):
+    """
+    A :class:`tests.utils.span.TestSpan` which is used as part of a span tree.
+
+    Each :class:`tests.utils.span.TestSpanNode` represents the current :class:`ddtrace.span.Span`
+    along with any children who have that span as it's parent.
+
+    This class can be used to assert on the parent/child relationships between spans.
+
+    Example::
+
+        class TestCase(BaseTestCase):
+            def test_case(self):
+                # TODO: Create spans
+
+                self.assert_structure( ... )
+
+                tree = self.get_root_span()
+
+                # Find the first child of the root span with the matching name
+                request = tree.find_span(name='requests.request')
+
+                # Assert the parent/child relationship of this `request` span
+                request.assert_structure( ... )
+    """
+    def __init__(self, root, children=None):
+        super(TestSpanNode, self).__init__(root)
+        object.__setattr__(self, '_children', children or [])
+
+    def get_spans(self):
+        """required subclass property, returns this spans children"""
+        return self._children
+
+    def assert_structure(self, root, children):
```

Comment:
I am not 100% sure here, I think the point of this function is to make assertions about the overall structure of a trace rather than about a single span.

The support for `children=None` was my way of allowing someone to determine how deep they wanted to test (e.g. the first 2-3 levels of spans, rather than the entire trace).

By not providing a default you force people to make that decision, and I don't want people to get used to not supplying `children` as a way to test that there are no children, since `children=None` ignores children testing and `children=()` asserts that there are no children.

Maybe we need to rethink the API for this helper in general, maybe there is a more ergonomic way to make this type of assertion. This really was only my first pass through this anyways, wouldn't be surprised if we could come up with something better.

Reply:
Ok, I updated as following:

```rst
        Example::

            def test_case(self):
                # Assert the following structure
                #
                # One root_span, with two child_spans, one with a requests.request span
                #
                # |                  root_span                |
                # |       child_span       | |   child_span   |
                # | requests.request |
                self.assert_structure(
                    # Root span with two child_span spans
                    dict(name='root_span'),

                    (
                        # Child span with one child of it's own
                        (
                            dict(name='child_span'),

                            # One requests.request span with no children
                            (
                                dict(name='requests.request'),
                            ),
                        ),

                        # Child span with no children
                        dict(name='child_span'),
                    ),
                )
```

The idea being that when you supply a list of children, each child can either be a `dict` or a `(dict(), ())`, if only a `dict` is provided then we assume `(dict(), ())`.

This way you don't always have to supply a list of children when none are expected.

Always assuming an empty list for children by default is great, to help ensure we don't skip any assertions unless the user specifically asks for that. We want to try and encourage asserting on a full trace if we can.

Reply:
I had an idea.

```python
NO_CHILDREN = object()

def assert_structure(root, children=NO_CHILDREN):
    if children is NO_CHILDREN:
        children = ()
```


Not sure if allowing `children` to be option adds more confusion to the top level API?

Reply:
Yep all makes sense! I think the approach is fine. I like the idea of encouraging/forcing the structure entirely but the test should do this anyway (if a child span is not included in the given span structure then fail).

Interesting thought about the depth. I could imagine cases where you'd only want to test down to a certain depth (multiple integrations enabled for instance).